examples/pennylane_run_rotoselect.py

```diff
@@ -0,0 +1,432 @@
+        params[d] += 2*np.pi
+    return cost(params, generators)
+
+
+def optimal_theta_and_gen_helper(d, params, generators, cost):
+    params[d] = 0.0
+    M_0 = cost(params, generators)  # M_0 independent of generator selection
+    for generator in ["X", "Y", "Z"]:
+        generators[d] = generator
+        params_cost = rotosolve(d, params, generators, cost, M_0)
+        # initialize optimal generator with first item in list, "X", and update if necessary
+        if generator == "X" or params_cost <= params_opt_cost:
+            params_opt_d = params[d]
+            params_opt_cost = params_cost
+            generators_opt_d = generator
+    return params_opt_d, generators_opt_d
+
+
+def rotoselect_cycle(cost, params, generators):
+    for d in range(len(params)):
+        params[d], generators[d] = optimal_theta_and_gen_helper(d, params, generators, cost)
+    return params, generators
+
+
+##############################################################################
+# Optimizing the circuit structure
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# We perform the optimization and print the optimal generators for the rotation gates. The minimized cost function
+# obtained by optimizing using Rotoselect is less than the minimized cost function from gradient descent or
+# Rotosolve, which were performed on the original circuit structure ansatz.
+# In other words, Rotoselect performs better without
+# increasing the depth of the circuit by selecting better gates for the task of minimizing the cost function.
+
+costs_rotoselect = []
+params = init_params.copy()
+init_generators = ["X", "Y"]
+generators = init_generators
+for _ in range(n_steps):
+    costs_rotoselect.append(cost(params, generators))
+    params, generators = rotoselect_cycle(cost, params, generators)
+
+print("Optimal generators are: {}".format(generators))
+
+# plot cost function vs. steps comparison
+fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(7, 3))
+plt.subplot(1, 2, 1)
+plt.plot(steps, costs_grad_desc, "o-")
+plt.title("grad. desc. on original ansatz")
+plt.xlabel("steps")
+plt.ylabel("cost")
+plt.subplot(1, 2, 2)
+plt.plot(steps, costs_rotoselect, "o-")
+plt.title("rotoselect")
+plt.xlabel("cycles")
+plt.ylabel("cost")
+plt.yticks(np.arange(-1.25, 0.80, 0.25))
+plt.tight_layout()
+plt.show()
+
+
+##############################################################################
+# Cost function surface for learned circuit structure
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# .. figure:: ../../examples/figures/learned_structure.png
+#    :scale: 65%
+#    :align: center
+#    :alt: learned_structure
+#
+# |
+# Finally, we plot the cost function surface for the newly discovered optimized
+# circuit structure shown in the figure above. It is apparent from the minima in the plot that
+# the new circuit structure is better suited for the problem.
+
+fig = plt.figure(figsize=(6, 4))
+ax = fig.gca(projection="3d")
+
+X = np.linspace(-4.0, 4.0, 20)
```

Comment:
let's bump up the resolution on both `X` and `Y` a little bit to make surface smoother

Reply:
Done